File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/sessions/handlers/SubscribeDeliverySessionHandler.java
package systems.zlink.samples.deliverydispatch.server.customergateway.sessions.handlers;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkTypedSessionPacketHandler;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActorDirectory;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SubscribeDeliverySessionHandler
        implements ZLinkTypedSessionPacketHandler<
                ZLinkSessionContext, Messages.SubscribeDeliveryReq> {
    private static final String CustomerId = "customer-1";

    private final ZLinkActorManager actors;
    private final CustomerActorDirectory customers;

    public SubscribeDeliverySessionHandler(
            ZLinkActorManager actors, CustomerActorDirectory customers) {
        this.actors = actors;
        this.customers = customers;
    }

    @Override
    public Class<Messages.SubscribeDeliveryReq> messageType() {
        return Messages.SubscribeDeliveryReq.class;
    }

    @Override
    public CompletionStage<Void> handle(
            ZLinkSessionContext context,
            ZLinkSessionDispatchContext dispatch,
            Messages.SubscribeDeliveryReq request) {
        Messages.FindCustomerActorReq find = new Messages.FindCustomerActorReq(CustomerId);
        return actors.find(find.customerId())
                .thenCompose(
                        found ->
                                found.isPresent()
                                        ? CompletableFuture.completedFuture(found.orElseThrow())
                                        : actors.getOrCreate(
                                                        CustomerId, SampleNames.CustomerActorType)
                                                .request(
                                                        new Messages.EnsureCustomerActorReq(
                                                                CustomerId))
                                                .submit()
                                                .thenApply(
                                                        SubscribeDeliverySessionHandler
                                                                ::requireActor))
                .thenCompose(
                        actor -> {
                            boolean requiresBinding =
                                    context.actors().find(actor.actorId()).isEmpty();
                            CompletionStage<ActorRef> bound =
                                    requiresBinding
                                            ? context.actors()
                                                    .bind(actor)
                                                    .thenApply(ignored -> actor)
                                            : CompletableFuture.completedFuture(actor);
                            return bound.thenApply(ignored -> requiresBinding);
                        })
                .thenAccept(
                        bound -> {
                            if (bound) {
                                System.out.println(
                                        "deliverydispatch-customer bound customer=" + CustomerId);
                            }
                            customers.subscribe(CustomerId, request.deliveryId());
                            context.client()
                                    .reply(new Messages.SubscribeDeliveryRes(request.deliveryId()))
                                    .submit();
                        });
    }

    private static ActorRef requireActor(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        throw new IllegalStateException("Customer actor creation was rejected.");
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/sessions/handlers/SubscribeDeliverySessionHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.sessions.handlers

import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.kotlin.ZLinkSuspendingTypedSessionPacketHandler
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActorDirectory
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.EnsureCustomerActorReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.FindCustomerActorReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.SubscribeDeliveryRes

class SubscribeDeliverySessionHandler(
    private val actors: ZLinkActorManager,
    private val customers: CustomerActorDirectory,
) : ZLinkSuspendingTypedSessionPacketHandler<ZLinkSessionContext, SubscribeDeliveryReq> {
    override fun packetName(): String = "SubscribeDeliveryReq"

    override fun messageType(): Class<SubscribeDeliveryReq> = SubscribeDeliveryReq::class.java

    override suspend fun handle(
        context: ZLinkSessionContext,
        dispatch: ZLinkSessionDispatchContext,
        message: SubscribeDeliveryReq,
    ) {
        val find = FindCustomerActorReq(CustomerId)
        val actor =
            actors.find(find.customerId).await().orElse(null)
                ?: actors
                    .kotlin()
                    .getOrCreate(CustomerId, SampleNames.CustomerActorType)
                    .request(EnsureCustomerActorReq(CustomerId))
                    .await()
                    .let(::requireActor)
        if (context.actors().find(actor.actorId()).isEmpty) {
            context.actors().bind(actor).await()
            println("deliverydispatch-customer bound customer=$CustomerId")
        }
        customers.subscribe(CustomerId, message.deliveryId)
        context.client().reply(SubscribeDeliveryRes(message.deliveryId)).submit()
    }

    private companion object {
        const val CustomerId = "customer-1"
    }

    private fun requireActor(result: ZLinkActorCreateResult): ActorRef =
        when (result) {
            is ZLinkActorCreateResult.Existing -> result.actor
            is ZLinkActorCreateResult.Created -> result.actor
            is ZLinkActorCreateResult.Rejected ->
                throw IllegalStateException("Customer actor creation was rejected.")
        }
}
